### PR TITLE
fix(upgrade): skip upgrade path check

### PR DESCRIPTION
**Pull request by mantissahz**
#### Which issue(s) this PR fixes:

Issue longhorn/longhorn#7539

#### What this PR does / why we need it:
- Add v140-to-v15x path back
-  Add a new flag `FlagUpgradeVersionCheck: "upgrade-version-check"`
    to accept the command to skip the upgrade path check.

### CODE REVIEW

File: app/daemon.go [hunk 1]
```diff
@@ -81,6 +82,10 @@ func DaemonCmd() cli.Command {
 				Name:  FlagKubeConfig,
 				Usage: "Specify path to kube config (optional)",
 			},
+			cli.BoolFlag{
+				Name:  FlagUpgradeVersionCheck,
+				Usage: "Have a warning log and an event for unrecommended upgrade path",
```
Comment (innobead) [inline app/daemon.go:87]: Usage should be `Enforce version checking for upgrades. If disabled, there will be no requirement for the necessary upgrade source version`

File: upgrade/upgrade.go [hunk 2]
```diff
@@ -35,7 +39,7 @@
 	LeaseLockName = "longhorn-manager-upgrade-lock"
 )
 
-func Upgrade(kubeconfigPath, currentNodeID, managerImage string) error {
+func Upgrade(kubeconfigPath, currentNodeID, managerImage string, upgradeVersionCheck bool) error {
```
Comment (innobead) [inline upgrade/upgrade.go:42]: `enableUpgradeVersionCheck bool`

File: upgrade/upgrade.go [hunk 3]
```diff
@@ -67,7 +71,13 @@
...
 	}
 
-	if err := upgradeutil.CheckUpgradePathSupported(namespace, lhClient); err != nil {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events("")})
+	eventRecorder := eventBroadcaster.NewRecorder(scheme, corev1.EventSource{Component: "longhorn-upgrade"})
+
+	if err := upgradeutil.CheckUpgradePathSupported(namespace, lhClient, eventRecorder, upgradeVersionCheck); err != nil {
```
Comment (innobead) [inline upgrade/upgrade.go:80]: Rename `upgradeutil.CheckUpgradePath`

File: upgrade/util/util.go [hunk 4]
```diff
@@ -235,6 +239,15 @@ func checkLHUpgradePathSupported(namespace string, lhClient lhclientset.Interfac
 		return fmt.Errorf("failed to upgrade since upgrading version %v is not valid", meta.Version)
 	}
 
+	if !upgradeVersionCheck {
+		if semver.Compare(lhCurrentVersion, "v1.4.0") < 0 {
+			return fmt.Errorf("failed to upgrade since downgrading from %v to %v is not supported", lhCurrentVersion, meta.Version)
```
Comment (innobead) [inline upgrade/util/util.go:244]:
```go
fmt.Errorf("failed to skip the upgrade check for %s to upgrade to %s, as it only supports the current version larger than v1.4.0", lhCurrentVersion, meta.Version)
```

**Review by innobead — CHANGES_REQUESTED** [review]
In general, LGTM. Some comments need to fix.

File: app/daemon.go (hunk 1 above)
Comment (mantissahz) [reply app/daemon.go]: Modified. Thanks

File: upgrade/util/util.go (hunk 4 above)
Comment (mantissahz) [reply upgrade/util/util.go]: Modified. Thanks

File: upgrade/upgrade.go (hunk 2 above)
Comment (mantissahz) [reply upgrade/upgrade.go]: Modified. Thanks

**Review by innobead — APPROVED** [review]

**Pull request merged**